plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java

```diff
@@ -210,6 +211,11 @@ public void asyncBlobUpload(WriteContext writeContext, ActionListener<Void> comp
         }
     }
 
+    @Override
+    public CompletableFuture<ReadContext> asyncBlobDownload(String blobName, boolean forceSingleStream) throws IOException {
+        throw new UnsupportedOperationException("S3 BlobContainer currently does not support async blob downloads");
+    }
+
```

Comment:
Can we keep the interface consistent and use a call back listener instead?

Reply:
I was going to suggest that we add a backlog task for making writes compatible with `Future` instead. 
I think `Future` provides a reduction in a lot of boilerplate code as well as a native Java interface support, than relying on the listener mechanism, which predates complete support for async programming natively. 
It also makes the code a lot for forward looking and compatible with any new features that come in with `Future` support.

What do you think?

Reply:
+1 for consistency.  Given this is an internal API i think we are free to refactor this to use Futures after this lands.  Or, we refactor the existing api first before adding download.

Reply:
I thought chained futures on return execute in a blocking fashion once you do a `.join` or `.get`. 
I would love to chat a bit more on `Future` v/s `Listener` pattern. It would be great to hear a bit more here @Bukhtawar  - https://github.com/opensearch-project/OpenSearch/issues/9236

Reply:
As I understand it, the following implementations are functionally identical:

CompletableFuture-based:
```java
asyncBlobDownload("foo", false).whenComplete((result, error) -> {
    if (error != null) {
        handleError(error);
    } else {
        handleSuccess(result);
    }
});
```

ActionListener-based:
```java
asyncBlobDownload("foo", false, new ActionListener<ReadContext>() {
    @Override
    public void onResponse(ReadContext result) {
        handleSuccess(result);
    }

    @Override
    public void onFailure(Exception e) {
        handleError(e);
    }
});
```

I think there are benefits to using the built-in concurrency primitives like CompletableFuture. For example, on the upload path we [adapt a CompletableFuture to the Listener interface](https://github.com/opensearch-project/OpenSearch/blob/639d28ad46927f8fbb6c940bc7efd4e935528526/plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java#L198-L205), and in the end [create a latch](https://github.com/opensearch-project/OpenSearch/blob/c3acf47b4d643c3a3ab86dc3b07fe722ac6e4982/server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java#L254) to block on the result. This likely would have been simpler as a chain of composed CompletableFutures ending in a `.get()` or `.join()` call to block on the result.

However, there is definitely an argument for consistency. @kotwanikunal Is there a clear benefit to using a CompletableFuture in this case, or is it mostly a style question at this point?

Reply:
Thanks @andrross! I am updating the API as we speak to make it consistent with the upload API. We can have this discussion here or on #9236 . I will summarize either ways. 
I am planning on creating a backlog item for downloads and uploads before we make that API public if they are functionally equivalent.

IMHO the big benefit is reduction in a lot of boilerplate code + native chaining (even with an array of futures) - all out of the box.
As seen from the example above, it is a lot more functional and readable. We can also rely on any Java API additions which support Future/CompletableFuture out of the box - making it a lot more *Future*proof. 
Sorry for the pun :)

Reply:
I would like to clarify that an `ActionListener` in no way restricts chaining. For instance look at the [wrapper](https://github.com/opensearch-project/OpenSearch/blob/9f415ebe87c1e26cde4f59da8fc93764c0ffa27d/server/src/main/java/org/opensearch/action/ActionListener.java#L72) utilities that help with chaining.
ActionListener serves to be a common IO agnostic interface for any calls made across the service for instance any node to node communication like TransportService#[sendRequest](https://github.com/opensearch-project/OpenSearch/blob/8afb22a525f4728e90826eabaa74109a9e745182/server/src/main/java/org/opensearch/transport/TransportService.java#L807-L812) has a ResponseHandler very similar to `ActionListener` in addition to the blocking `TransportFuture` which finds very rare usage.
One of the reasons ActionListener is used extensively is because of the Netty framework that is non-blocking and most REST request modelled as a callback which doesn't require blocking on response but a callback that send the response over the channel once available.

Reply:
> One of the reasons ActionListener is used extensively

I think the main reason why ActionListener is used instead of CompletableFuture is that ActionListeners pre-date the introduction of CompletableFuture in Java 8. I think the question here is whether it is okay to start introducing CompletableFutures into new functionality (with the benefit of using a standard language primitive with all the built-in capabilities that go with it), or it is better to stay with ActionListener for consistency across the project.